pkg/storegateway/series_chunks.go

```diff
@@ -398,15 +424,60 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		nextSet.series = filterNativeHistogramChunks(nextSet.series)
+		if len(nextSet.series) == 0 {
+			nextSet.release()
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistogramChunks(seriesEntries []seriesEntry) []seriesEntry {
+	seriesWriteIdx := 0
+	for i, series := range seriesEntries {
+		chksWriteIdx := 0
+		for j, chk := range series.chks {
+			// The bucketChunkReader will skip populating native histogram chunks when ignoreNativeHistograms is true,
+			// leaving chk.Raw as nil.
+			if chk.Raw == nil {
+				continue
+			}
+
+			series.chks[chksWriteIdx] = series.chks[j]
+			if series.refs != nil {
+				// Not all callers will have series.refs set, but when it is set we expect refs and chks to have a
+				// one-to-one relationship.
+				series.refs[chksWriteIdx] = series.refs[j]
+			}
+
+			chksWriteIdx++
+		}
+		seriesEntries[i].chks = series.chks[:chksWriteIdx]
+
+		if len(seriesEntries[i].chks) == 0 {
+			continue
+		}
+		// We swap the two series entries so the entry that would've just been overwritten at `seriesWriteIdx` is still
+		// reachable in `seriesChunksSet.release`.
+		swp := seriesEntries[seriesWriteIdx]
+		seriesEntries[seriesWriteIdx] = seriesEntries[i]
+		seriesEntries[i] = swp
```

Comment:
this also works in go:
```suggestion
		seriesEntries[seriesWriteIdx], seriesEntries[i] = seriesEntries[i], seriesEntries[seriesWriteIdx]
```